Sources/BraintreeCore/RepeatingTimer.swift

```diff
@@ -0,0 +1,58 @@
+import Foundation
+
+final class RepeatingTimer {
+    
+    /// Amount of time, in seconds.
+    private let timeInterval: Int
+    
+    init(timeInterval: Int) {
+        self.timeInterval = timeInterval
+    }
+    
+    private lazy var timer: DispatchSourceTimer = {
+        let timerSource = DispatchSource.makeTimerSource(queue: DispatchQueue.main)
+        timerSource.schedule(
+            deadline: .now() + .seconds(timeInterval),
+            repeating: .seconds(timeInterval),
+            leeway: .seconds(1)
+        )
+        timerSource.setEventHandler(handler: { [weak self] in
+            self?.eventHandler?()
+        })
+        return timerSource
+    }()
+    
+    var eventHandler: (() -> Void)?
+    
+    private enum State {
+        case suspended
+        case resumed
+    }
+    
+    private var state: State = .suspended
+    
+    deinit {
+        timer.setEventHandler {}
+        timer.cancel()
+        /*
+         If the timer is suspended, calling cancel without resuming
+         triggers a crash. This is documented here https://forums.developer.apple.com/thread/15902
```

Comment:
Based on this comment this seems like it could be the main source of the crash. Have we tried just adding a `resume` call to our `deinit`?

Reply:
Yes, however, I think moving the logic to a separate class would help us segregate the responsibilities of each class. Another benefit is that we can better manage the instantiation of `DispatchSource` (RepeatingTimer). Previously, it was possible to instantiate it more than once, even when we [validated](https://github.com/braintree/braintree_ios/blob/f03147676b81f938c04c22525ffc60befd07441d/Sources/BraintreeCore/Analytics/BTAnalyticsService.swift#L121) that it was equal to nil.